torch/csrc/jit/passes/shape_analysis.cpp

```diff
@@ -201,10 +200,60 @@ class ShapePropagator {
         continue;
       }
     }
-
     return tensor_types;
   }
 
+  c10::ScalarType unionScalarTypes(c10::ScalarType original, c10::ScalarType next) {
+    if (original == c10::ScalarType::Undefined) {
+      return next;
+    } else {
+      return c10::promoteTypes(original, next);
+    }
+  }
+
+  // Promotes result types for arithmetic operations using new type promotion logic.
+  // See tensor_attributes.rst for details.
+  // This function handles the dimmed and zero-dim case, not the `wrapped-number`
```

Comment:
Nit: `wrapped-number` isn't a term within JIT. I would rephrase, something like "This function handles `add, sub, mul, div` when first two arguments are tensors. For the same ops where one input is a Scalar is handled separately.